service/storage_proxy.cc

```diff
@@ -973,7 +973,11 @@ static result<db::per_partition_rate_limit::info> choose_rate_limit_info(
     db::per_partition_rate_limit::account_and_enforce enforce_info{
         .random_variable = random_variable_for_rate_limit(),
     };
-    if (coordinator_in_replica_set && erm->get_sharder(*s).shard_of(token) == this_shard_id()) {
+    // It's fine to use shard_for_reads() because in case of no migration this is the
+    // shard used by all requests. During migration, it is the shard used for request routing
+    // by drivers during most of the migration. It changes after streaming, in which case we'll
+    // fall back to throttling on replica side, which is suboptimal but acceptable.
+    if (coordinator_in_replica_set && erm->shard_for_reads(*s, token) == this_shard_id()) {
```

Comment:
I see a small problem.

If we get into this `if` then we increment our counter and decide whether to accept or not based on that. If we accept, then we tell other replicas to increase their counters while they unconditionally accept (the `account_only` case). There is a piece of logic in `abstract_write_response_handler::apply_locally` which prevents the write from being accounted twice when applying locally (`adjust_rate_limit_for_local_operation`). If we are currently doing double writes (to both shards) then this piece of logic will "accidentally" cause the counter not to be increased locally on the other shard (the other one than `shard_for_reads`).

Perhaps this check could be somehow moved to mutate_locally? We would have to know whether we are the current coordinator and whether this was the coordinator shard or not. If it's a hassle then I suppose we can also ignore it as it shouldn't affect the count too much, but perhaps we should write this down as a fixme/issue/etc.

Reply:
What could be the consequences of that? The consequences of not counting on the other shard will be that once migration is finished and the coordinating shard transitions to the other one, we will allow more traffic for a while because the other shard didn't see the writes. It's similar to what happens when the traffic starts after an idle period, except that in this case the new shard is not idle. It may have full active queue of requests.

Reply:
Maybe we can ignore the problem, since we migrate only a fraction of tablets at a time, so impact is also reduced by that.

Reply:
In case of non-shard-aware writes, each replica shard independently decides whether to accept the write or not. The assumption is that the counter values on all replica are close together, so they will agree on the decision most of the time. If we only have non-shard-aware writes then it's fine because all replicas will increment the counters, but it's not the case for shard-aware writes because of what I described. If you mix shard-aware and non-shard-aware writes then the one problematic shard will accept more frequently than the others which might be a source of inconsistency.

This is a convoluted case because it requires all of the following to happen:

- Have shard-aware clients,
- Have non-shard-aware clients,
- Have a partition overloaded by writes which is also being currently migrated.

I agree that it's not the end of the world, but I think we should at least leave a fixme or create an issue not to forget about it.

Reply:
I will create an issue once this is merged. I think it's better than leaving an elaborate FIXME in a random place.